### PR TITLE
Adding support for table and column renaming using Regex 

### PR DESCRIPTION
**Pull request by evandromendonca**
Hello, first I'd like to thank you for this tool, it has been very useful and I am more than happy to contribute.

I had the need to remove some prefixes from my tables. I found out [this](https://github.com/ErikEJ/EFCorePowerTools/issues/251) thread but didn't want to map my entire database.

So I took inspiration from there and included 4 new properties in the Schema object. These will receive the pattern to be matched and a string to replace tables and columns names.

**TableRegexPattern:** pattern to be matched by table names
**TablePatternReplaceWith:** string that will replace the occurrences of the pattern in tables names
**ColumnRegexPattern:** pattern to be matched by column names
**ColumnPatternReplaceWith:** string that will replace the occurrences of the pattern in columns names

The regex renaming also have lower priority over the current renaming method, and therefore, if a table is renamed using _"NewName"_, the NewName method will prevail over the regex

Example of use: 

efpt.renaming.json
```
[
  {
    "UseSchemaName": false,
    "SchemaName": "dbo",
    "TableRegexPattern": "^Old",
    "TablePatternReplaceWith": "New",
    "ColumnRegexPattern": "^Old",
    "ColumnPatternReplaceWith": "New",
    "Tables": [
      {
        "Name": "course_sched",
        "NewName":  "CourseOffering",
        "Columns": [
          {
            "Name": "subj_area_cd",
            "NewName": "SubjectArea"
          }
        ]
      }
    ]
  }
]
```

### CODE REVIEW

**Comment by ErikEJ** [comment]
You must rebase on latest master

**Comment by ErikEJ** [comment]
Would it be possible to design for expetions from the general rules for individual tables and/or columns?

**Comment by evandromendonca** [comment]
> Would it be possible to design for expetions from the general rules for individual tables and/or columns?

Sorry, I didn't understand what you meant there.

**Comment by ErikEJ** [comment]
I just didn't read this properly:

The regex renaming also have lower priority over the current renaming method, and therefore, if a table is renamed using "NewName", the NewName method will prevail over the regex.

File: src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs [hunk 1]
```diff
@@ -42,28 +43,32 @@ public override string GenerateCandidateIdentifier(DatabaseTable originalTable)
...
-                candidateStringBuilder.Append(newTableName);
-
-                return candidateStringBuilder.ToString();
+            if (schema.Tables.Any(t => t.Name == originalTable.Name))
+            {
+                newTableName = schema.Tables.FirstOrDefault(t => t.Name == originalTable.Name)?.NewName;
+            }
+            else if (!string.IsNullOrEmpty(schema.TableRegexPattern) && schema.TablePatternReplaceWith != null)
+            {
+                newTableName = Regex.Replace(originalTable.Name, schema.TableRegexPattern, schema.TablePatternReplaceWith);
```
Comment (ErikEJ) [inline src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs:55]: Please add options with a timeout

File: src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs [hunk 2]
```diff
@@ -96,6 +101,17 @@ public override string GenerateCandidateIdentifier(DatabaseColumn originalColumn
                 }
             }
 
+            if (!string.IsNullOrEmpty(schema.ColumnRegexPattern) && schema.ColumnPatternReplaceWith != null)
+            {
+                string newColumnName = Regex.Replace(originalColumn.Name, schema.ColumnRegexPattern, schema.ColumnPatternReplaceWith);
```
Comment (ErikEJ) [inline src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs:106]: Please add options with a timeout

File: src/GUI/RevEng.Shared/Schema.cs [hunk 3]
```diff
@@ -13,5 +13,13 @@ public Schema()
...
+
+        public string TableRegexPattern { get; set; }
+
+        public string TablePatternReplaceWith { get; set; }
+
+        public string ColumnRegexPattern { get; set; }
+
+        public string ColumnPatternReplaceWith { get; set; }
     }
 }
```
Comment (ErikEJ) [inline src/GUI/RevEng.Shared/Schema.cs:25]: Remove the .zip file for your PR - I will update when this has been merged

**Comment by ErikEJ** [comment]
Please address review comments

File: src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs (hunk 1 above)
Comment (evandromendonca) [reply src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs]: Did add a timeout to regex replace method

File: src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs (hunk 2 above)
Comment (evandromendonca) [reply src/GUI/RevEng.Core/ReplacingCandidateNamingService.cs]: Did add a timeout to regex rename method

**Comment by ErikEJ** [comment]
Great, thanks a lot for the contribution

**Comment by jamshally** [comment]
This is a great feature. Thanks!

**Comment by ErikEJ** [comment]
You are welcome, agree that it rounds out the renaming feature nicely.

**Comment by engineruguz** [comment]
I'm sorry but I don't understand how to use this feature.

i add

    "TableRegexPattern": "^.+",
    "TablePatternReplaceWith": "$0Entity"

to the efpt.config.json file but it gets overwritten when i rerun the reverse engineer.

**Comment by ErikEJ** [comment]
@engineruguz No worries - You got the file name wrong - see https://github.com/ErikEJ/EFCorePowerTools/wiki/Reverse-Engineering#custom-renaming-with-efptrenamingjson

**Comment by engineruguz** [comment]
oooh! i see. feel dumb lol

**Comment by engineruguz** [comment]
hi! sorry. i still can't find out why it's not working? i created efpt.renaming.json file at project root like below.

```
{
  "UseSchemaName": false,
  "SchemaName": "dbo",
  "TableRegexPattern": "^.+$",
  "TablePatternReplaceWith": "$0Entity"
}
```

i'm trying to add Entity at the end of every table. not sure about $0 is the correct syntax though but not getting anything different.

**Comment by ErikEJ** [comment]
Wrong file format...

**Comment by engineruguz** [comment]
uh ofc! sorry. thanks

**Comment by marcziss** [comment]
Either I'm doing something very dumb or there's an issue with ColumnPatternReplaceWith

   {
      "ColumnPatternReplaceWith": "TEST_COLUMN_",
      "ColumnRegexPattern": "^",
      "SchemaName": "PDM",
      "TablePatternReplaceWith": "TEST_TABLE_",
      "TableRegexPattern": "^",
      "UseSchemaName": false
   }

In this example using .net 5 (and latest version of vsix installed today) Every table is renamed properly with the TEST_TABLE_ prefix, however no column is renamed at all. Is there something I'm missing?

**Comment by ErikEJ** [comment]
@marcziss if you have an issue, please create one with detailed repro steps.

**Comment by HansB159** [comment]
@marcziss I struggled with the same problem. Solution: add "Table: []". That magically worked for me.

**Pull request merged**